src/Parser.php

```diff
@@ -0,0 +1,340 @@
+<?php
+
+namespace Phlib\Mail;
+
+use Phlib\Mail\Exception\RuntimeException;
+
+class Parser
+{
+    /**
+     * @var bool
+     */
+    private $isFile;
+
+    /**
+     * @var string
+     */
+    private $source;
+
+    /**
+     * @var resource
+     */
+    private $mimeMail;
+
+    /**
+     * @var array
+     */
+    private $structure = [];
+
+    /**
+     * @param boolean $isFile
+     * @param string $source
+     */
+    public function __construct($isFile, $source)
```

Comment:
Not keen on this as a constructor, especially `isFile`. These seem like factory methods. The 'Parser' name also suggests a stateless type. `$parser->parse('...');` like `$sanitiser->sanitise();` etc.

Maybe, `$mail = (new Parser)->parseFromFile($filename);` or `$mail = (new Parser)->parse($source);`.

Reply:
Then I think that just puts us back to where we already were (which is perhaps the correct thing!) What I was trying to aim for was a single class which would have knowledge of the `mailparse_*` methods, and be able to correctly clean up once the parsing is complete. The `__destruct()` seemed a nice way of achieving this, but couldn't be used on a factory-style approach as the *Factory* would not be destroyed once parsing is complete.

@Pudge601 's other suggestion was a `try ... finally` for both of the `fromString()` and `fromFile()` methods, which would perform the clean-up after returning the newly constructed *Mail* object.

Reply:
Far more reasonable to have a stateful 'parser' than a stateful 'factory' (as it is currently).

At some point something needs to be stateful to hold the `mimeMail` resource and the structure - either that or we refactor it to pass those two things around into all the method (even recursively for the case of `parsePart`, which just seems nasty).

Calling it a parser makes sense because that's what it's doing (again, better than the current 'factory', which is doing a lot more than just creating instances of things).

You could store the state of the current parsing context in a `ParsingContext` class (or something similarly named), but then you either pass that around everywhere, or move the logic for the parsing to the parsing context class itself; at which point you basically have exactly what we've come up with here, except it's named `ParsingContext` instead of `Parser` (and considering it would be doing the parsing, it seems reasonable to call it a `Parser`).